src/pip/_internal/cmdoptions.py

```diff
@@ -60,6 +61,42 @@ def getname(n):
         )
 
 
+def check_dist_restriction(options, check_target=False):
+    """Function for determinming if custom platform options are allowed.
+
+    :param options: The OptionParser options.
+    :param check_target: Whether or not to check if --target is being used.
+    """
+    dist_restriction_set = any([
+        options.python_version,
+        options.platform,
+        options.abi,
+        options.implementation,
+    ])
+
+    binary_only = FormatControl(set(), {':all:'})
+    no_sdist_dependencies = (
+        options.format_control != binary_only and
+        not options.ignore_dependencies
+    )
+
+    if dist_restriction_set and no_sdist_dependencies:
```

Comment:
I know it wasn't there originally but adding a comment that explains why these error messages are raised would be appreciated.